packages/block-library/src/site-logo/index.php

```diff
@@ -71,56 +71,83 @@ function register_block_core_site_logo() {
- *
- * @param string $custom_logo The custom logo set by a theme.
- *
- * @return string The site logo if set.
+ * Expose the custom_logo theme-mod in the settings REST API.
  */
-function override_custom_logo_theme_mod( $custom_logo ) {
-	$site_logo = get_option( 'site_logo' );
-	return false === $site_logo ? $custom_logo : $site_logo;
-}
+register_setting(
+	'general',
+	'theme_mods_' . get_option( 'stylesheet' ),
+	array(
+		'type'         => 'object',
+		'show_in_rest' => array(
+			'name'   => 'theme_mods_' . get_option( 'stylesheet' ),
+			'schema' => array(
+				'type'       => 'object',
+				'properties' => array(
+					'custom_logo' => array( 'type' => 'integer' ),
+				),
+			),
+		),
+	)
+);
 
 /**
- * Syncs the site logo with the theme modified logo.
+ * Expose the "stylesheet" setting in the REST API.
+ */
+register_setting(
+	'general',
+	'stylesheet',
+	array(
+		'type'         => 'string',
+		'show_in_rest' => true,
+	)
+);
+
+/**
+ * Filters the value of a setting recognized by the REST API.
  *
- * @param string $custom_logo The custom logo set by a theme.
+ * Hijacks the value for custom_logo theme-mod.
  *
- * @return string The custom logo.
+ * @param mixed  $result Value to use for the requested setting. Can be a scalar
+ *                       matching the registered schema for the setting, or null to
+ *                       follow the default get_option() behavior.
+ * @param string $name   Setting name (as shown in REST API responses).
+ *
+ * @return null|array
  */
-function sync_site_logo_to_theme_mod( $custom_logo ) {
-	// Delete the option when the custom logo does not exist or was removed.
-	// This step ensures the option stays in sync.
-	if ( empty( $custom_logo ) ) {
-		delete_option( 'site_logo' );
-	} else {
-		update_option( 'site_logo', $custom_logo );
+function gutenberg_rest_pre_get_setting_filter_custom_logo( $result, $name ) {
+	if ( 'theme_mods_' . get_option( 'stylesheet' ) === $name ) {
+		return array(
+			'custom_logo' => get_theme_mod( 'custom_logo' ),
+		);
 	}
-	return $custom_logo;
 }
+add_filter( 'rest_pre_get_setting', 'gutenberg_rest_pre_get_setting_filter_custom_logo', 10, 2 );
 
 /**
- * Register a core site setting for a site logo
+ * Filters whether to preempt a setting value update via the REST API.
+ *
+ * Hijacks the saving method for theme-mods.
+ *
+ * @param bool   $result Whether to override the default behavior for updating the
+ *                       value of a setting.
+ * @param string $name   Setting name (as shown in REST API responses).
+ * @param mixed  $value  Updated setting value.
+ *
+ * @return bool
  */
-function register_block_core_site_logo_setting() {
-	register_setting(
-		'general',
-		'site_logo',
-		array(
-			'show_in_rest' => array(
-				'name' => 'site_logo',
-			),
-			'type'         => 'integer',
-			'description'  => __( 'Site logo.' ),
-		)
-	);
+function gutenberg_rest_pre_set_setting_filter_theme_mods( $result, $name, $value ) {
```

Comment:
This function is used so that the theme-mods we set get merged with the pre-existing ones instead of replacing them. Without this, after saving the logo we'd only have the `custom_logo` theme-mod in the database and everything else would be getting trashed.